src/allmydata/test/test_storage_https.py

```diff
@@ -0,0 +1,232 @@
+    """
+
+    def to_file(self, key_or_cert) -> str:
+        """
+        Write the given key or cert to a temporary file on disk, return the
+        path.
+        """
+        path = self.mktemp()
+        with open(path, "wb") as f:
+            if isinstance(key_or_cert, x509.Certificate):
+                data = key_or_cert.public_bytes(serialization.Encoding.PEM)
+            else:
+                data = key_or_cert.private_bytes(
+                    encoding=serialization.Encoding.PEM,
+                    format=serialization.PrivateFormat.TraditionalOpenSSL,
+                    encryption_algorithm=serialization.NoEncryption(),
+                )
+            f.write(data)
+        return path
+
+    def generate_private_key(self):
+        """Create a RSA private key."""
+        return rsa.generate_private_key(public_exponent=65537, key_size=2048)
+
+    def generate_certificate(
+        self, private_key, expires_days: int = 10, org_name: str = "Yoyodyne"
+    ):
+        """Generate a certificate from a RSA private key."""
+        subject = issuer = x509.Name(
+            [x509.NameAttribute(NameOID.ORGANIZATION_NAME, org_name)]
+        )
+        expires = datetime.datetime.utcnow() + datetime.timedelta(days=expires_days)
+        return (
+            x509.CertificateBuilder()
+            .subject_name(subject)
+            .issuer_name(issuer)
+            .public_key(private_key.public_key())
+            .serial_number(x509.random_serial_number())
+            .not_valid_before(min(datetime.datetime.utcnow(), expires))
+            .not_valid_after(expires)
+            .add_extension(
+                x509.SubjectAlternativeName([x509.DNSName("localhost")]),
+                critical=False,
+                # Sign our certificate with our private key
+            )
+            .sign(private_key, hashes.SHA256())
+        )
+
+    @asynccontextmanager
+    async def listen(self, private_key_path, cert_path):
+        """
+        Context manager that runs a HTTPS server with the given private key
+        and certificate.
+
+        Returns a URL that will connect to the server.
+        """
+        endpoint = serverFromString(
+            reactor,
+            "ssl:privateKey={}:certKey={}:port=0:interface=127.0.0.1".format(
+                quoteStringArgument(str(private_key_path)),
+                quoteStringArgument(str(cert_path)),
+            ),
+        )
+        root = Data(b"YOYODYNE", "text/plain")
+        root.isLeaf = True
+        listening_port = await endpoint.listen(Site(root))
+        try:
+            yield f"https://127.0.0.1:{listening_port.getHost().port}/"
+        finally:
+            await listening_port.stopListening()
+            # Make sure all server connections are closed :( No idea why this
+            # is necessary when it's not for IStorageServer HTTPS tests.
+            await deferLater(reactor, 0.001)
```

Comment:
My guess is that it's necessary because even with non-persistent connections, `Agent` still fires application-code-facing Deferreds before it finishes closing its connections.  It's probably necessary for IStorageServer HTTPS tests too but they get lucky more often (or they got persistently lucky once and some other aspect of cleanup lets the reactor tick just enough for those connections to end).

I have a similar work-around in a test fixture managing a Treq instance - https://github.com/PrivateStorageio/ZKAPAuthorizer/blob/main/src/_zkapauthorizer/tests/fixtures.py#L195-L205

:(

Reply:
OK.

Reply:
I filed https://tahoe-lafs.org/trac/tahoe-lafs/ticket/3887 which we could link to from here.